pkg/controller/datavolume-controller.go

```diff
@@ -2559,10 +2604,22 @@ func getDefaultAccessModes(c client.Client, storageClass *storagev1.StorageClass
 		return nil, errors.Wrap(err, "no accessMode defined on DV, cannot get StorageProfile")
 	}
 
-	if len(storageProfile.Status.ClaimPropertySets) > 0 &&
-		len(storageProfile.Status.ClaimPropertySets[0].AccessModes) > 0 {
-		accessModes := storageProfile.Status.ClaimPropertySets[0].AccessModes
-		return accessModes, nil
+	if len(storageProfile.Status.ClaimPropertySets) > 0 {
+		// check for access modes matching with given pvc volume mode
+		defaultAccessModes := []corev1.PersistentVolumeAccessMode{}
+		for _, cps := range storageProfile.Status.ClaimPropertySets {
+			if cps.VolumeMode != nil && *cps.VolumeMode == *pvcVolumeMode {
+				if len(cps.AccessModes) > 0 {
+					return cps.AccessModes, nil
+				}
+			} else if len(cps.AccessModes) > 0 && len(defaultAccessModes) == 0 {
+				defaultAccessModes = cps.AccessModes
```

Comment:
so default can be the last accessModes found?
why not take first one as in VolumeModes  `storageProfile.Status.ClaimPropertySets[0].AccessModes`

Reply:
no the default will be the first access modes found. I didnt just take storageProfile.Status.ClaimPropertySets[0].AccessModes cause I thought it is possible that it will be an empty list.